### ISSUE TITLE
[TEST] New ZIM files without file extension in the URL

### ISSUE THREAD

**Issue by kelson42** [kiwix/kiwix-js#475]
`mwoffliner` 1.8 dev works carries on and we have the first ZIM file generated. A few things have changed in the ZIM file but maybe the most important is that the URL don't have anymore the `.html` file extension for articles. ZIM URL paths should be now exactly the same as the URL path of the online Wikipedia.

Please have a look to an example of ZIM file http://tmp.kiwix.org/bm_wikipedia.libzim.zim and let us know if there is a problem somewhere with Kiwix JS.

**Comment by Jaifroid** [comment]
@kelson42 On a quick test, hyperlinks and searching for articles work fine in Kiwix JS and Kiwix JS Windows. I note that the dirEntry.title field is missing, for most articles in this ZIM, and Kiwix JS therefore uses the dirEntry.url to match a search term. But I think that's a problem this ZIM rather than with Kiwix JS -- the title field shouldn't be blank in articles like `Afrika` and `Senegal`, should it?

@mossroy The lines that substitute url for title are lines 705+ of app.js, which is why in this ZIM, searching for "S" produces the list shown in the screenshot below. This shows that all of those entries with square brackets are erroneously missing the dirEntry.title attribute:

```
        for (var i = 0; i < listLength; i++) {
            var dirEntry = dirEntryArray[i];
            var title = dirEntry.title ? dirEntry.title : '[' + dirEntry.url + ']';
            articleListDivHtml += '<a href="#" dirEntryId="' + dirEntry.toStringId().replace(/'/g, '&apos;') +
                '" class="list-group-item">' + title + '</a>';
        } 
```
![image](https://user-images.githubusercontent.com/4304337/53283390-8ecf6500-373d-11e9-8533-40730293d6e3.png)

**Comment by kelson42** [comment]
@Jaifroid The title should (or is) here, see:
```
$ zimdump -L bm_wikipedia.libzim.zim | grep text/html | more
A	(ߒߞߏ)	(ߒߞߏ)	786	A	text/html	4812
A	7000_km_plus_loin	7000_km_plus_loin	787	A	text/html	3432
A	ARP	ARP	788	A	text/html	3766
A	A_Banna	A_Banna	789	A	text/html	3396
A	Abbaretz	Abbaretz	790	A	text/html	4889
A	Aerospacial	Aerospacial	795	A	text/html	3126
A	Afel_Bokum	Afel_Bokum	797	A	text/html	3129
A	Afrika	Afrika	800	A	text/html	10093
A	Agnes_Monica	Agnes_Monica	801	A	text/html	3329
A	Aigrefeuille-sur-Maine	Aigrefeuille-sur-Maine	803	A	text/html	4959
A	Aisland	Aisland	804	A	text/html	5322
A	Akra	Akra	805	A	text/html	3374
A	Albert_Einstein	Albert_Einstein	806	A	text/html	5015
A	Albuquerque	Albuquerque	807	A	text/html	3453
A	Alfa_Umar_Konare	Alfa_Umar_Konare	808	A	text/html	4763
A	Alger	Alger	809	A	text/html	4287
A	Ali_Baba_ni_n'sonke_binaani_kɛlɛ	Ali_Baba_ni_n'sonke_binaani_kɛlɛ	813	A	text/html	5560
A	Ali_Farka_Ture	Ali_Farka_Ture	816	A	text/html	10212
A	Alimankɛlɛ_fɔlɔ	Alimankɛlɛ_fɔlɔ	817	A	text/html	3385
A	Alimaɲi	Alimaɲi	819	A	text/html	7265
A	Alizeri	Alizeri	820	A	text/html	12533
A	Amadu_Tumani_Ture	Amadu_Tumani_Ture	824	A	text/html	3776
A	Amadu_ni_Mariyam	Amadu_ni_Mariyam	825	A	text/html	5738
A	Ambelau	Ambelau	826	A	text/html	3373
A	Amerika_ka_Kelenyalen_Jamanaw	Amerika_ka_Kelenyalen_Jamanaw	828	A	text/html	6844
A	Amsterdam	Amsterdam	829	A	text/html	3364
A	An_Be_Nɔ_do	An_Be_Nɔ_do	830	A	text/html	3419
A	Anaheim	Anaheim	831	A	text/html	3431
A	Ancenis	Ancenis	832	A	text/html	4808
A	Anchorage	Anchorage	833	A	text/html	3456
A	Anetz	Anetz	834	A	text/html	4646
A	Anggun	Anggun	835	A	text/html	3355
A	Angilɛkan	Angilɛkan	836	A	text/html	4272
A	Angola	Angola	837	A	text/html	11626
A	Ankara	Ankara	838	A	text/html	3362
A	Antibes	Antibes	842	A	text/html	4850
A	Antony	Antony	843	A	text/html	5571
A	Antropoloji	Antropoloji	845	A	text/html	4144
A	Arcachon	Arcachon	846	A	text/html	4837
A	Argentina	Argentina	847	A	text/html	5274
A	Arlington_(Texas)	Arlington_(Texas)	848	A	text/html	3475
A	Asia	Asia	849	A	text/html	3626
A	Asnières-sur-Seine	Asnières-sur-Seine	850	A	text/html	4940
A	Athens	Athens	851	A	text/html	3341
A	Atlanta	Atlanta	852	A	text/html	3431
A	Atlantik_Kogoji	Atlantik_Kogoji	853	A	text/html	3725
A	Austin_(Texas)	Austin_(Texas)	854	A	text/html	3454
A	Avessac	Avessac	855	A	text/html	4540
A	Ayiti	Ayiti	857	A	text/html	4882
```
So, I wonder about this comment!

**Comment by Jaifroid** [comment]
Hmm, thanks for that @kelson42. Compare the same dirEntry dump produced by Kiwix JS (screenshot below). Our dump includes redirects, whereas yours is limited to text/html, and it is those for which Kiwix JS seems unable to find the dirEntry.title field. @mossroy have you any idea what's going on here?

![image](https://user-images.githubusercontent.com/4304337/53284766-8a607780-3750-11e9-94e4-7a8d0e1bb635.png)

**Comment by Jaifroid** [comment]
Actually, @kelson42 , looking at your list, it seems that your zimdump executable is "filling in" the blank title fields with the URL. Witness the fact that many of your titles have got underscores, e.g. `Afel_Bokum` or `Albert_Einstein`. Those have been filled in from the dirEntry.url field by zimdump. The underlying title should clearly be `Afel Bokum` (with a space, not underscore) and `Albert Einstein`. Titles shouldn't have underscores instead of spaces. So, I suspect it is a problem with the ZIM, not our way of reading the title field.

**Comment by Jaifroid** [comment]
PS We did investigate a previous problem with missing dirEntry.title fields for very specific types of files in the A namespace in some TEDx ZIMs. At the time, @mossroy investigated the code in libzim and found that you have code that fills in the title field from the url if the title is missing. The code is here:   

https://github.com/openzim/libzim/blob/684c355a6c56208c3d9e0be7990c4923b7ed8c50/src/_dirent.h#L84

and the specific line that makes the substitution is 

`const std::string& getTitle() const     { return title.empty() ? url : title; }`

This is the reason why your zimdump (which I imagine uses libzim) is showing a title field when in fact it is blank. We chose to put square brackets in such cases. This strongly suggests to me that the problem is in mwoffliner, since it isn't populating the title field correctly, at least in this ZIM.

**Comment by kelson42** [comment]
@Jaifroid @mossroy Thank you for reporting the pb around the fake "title". I have open a ticket at the libzim title to stop faking it if absent, see https://github.com/openzim/libzim/issues/208
@ISNIT0 Any Idea why the article put through `node-libzim` seem to not have any title?

**Comment by ISNIT0** [comment]
Yes, I see that I'm not passing in a title parameter, I'll implement this now

**Comment by kelson42** [comment]
@ISNIT0 good, this should be only for the text/html stuff of course.

**Comment by mossroy** [comment]
I tested a bit in both jQuery and SW modes, and saw no issue.
It's not surprising because we never look for a ".html" suffix, or assume there is one.

I also confirm what @Jaifroid said about the title attribute : nice investigation.

It would be worth trying again when this title will be populated, but I see no reason why it would cause any regression.

**Comment by kelson42** [comment]
@mossroy We will upload a new version with the titles.

**Comment by ISNIT0** [comment]
Just noticed that some titles have HTML in them, I'll strip that then upload one

**Comment by ISNIT0** [comment]
This was built with latest master of `mwoffliner` https://framadrop.org/r/Y_Ab6yt-Hs#Q/BHkNiqSaJ6j6cUP1g7pnaHGBrxG0Q/WXFJlerC56I=

**Comment by mossroy** [comment]
It seems to work well, including the welcome page with large images.

Most articles have a title, but there are still a few ones that don't. At least : 
- Alexanderschlacht (that redirects to The_Battle_of_Alexander_at_Issus)
- HomePage, MAINPAGE, MP:, all the MP:* and a few other ones (that all redirect to Main_Page)
- Tower7 and WTC7 (that redirect to 7_World_Trade_Center)
All of them only have one word : could it be the reason?

**Comment by ISNIT0** [comment]
The "title" is set to be `articleId.replace(/_/g,'')` so no change if there are no underscores :)

**Comment by Jaifroid** [comment]
I confirm that this also works well in Kiwix JS Windows. I notice the same (very few) items that should have titles but don't. You can see these circled in screen shot below from the Kiwix JS Windows archive index page for the ZIM. All entires that consist of a single word are not getting titles.

I think your regex will fail if it doesn't find an underscore in the articleId. That's what's causign this issue IMHO.

![image](https://user-images.githubusercontent.com/4304337/53360058-7b362100-392c-11e9-9d10-7f5e8a9d39f3.png)

**Comment by mossroy** [comment]
For the articles mentioned above, kiwix-js detects an empty title (and puts the url inside brackets instead) : 
![image](https://user-images.githubusercontent.com/890923/53360411-c0128580-3935-11e9-80ae-77706f3b36a0.png)

If I put a breakpoint, the title is empty : 

![image](https://user-images.githubusercontent.com/890923/53360326-8a6d9c80-3935-11e9-8d63-454758571c9c.png)

**Comment by kelson42** [comment]
The title should obviously comes from the `<title>` DOM node (like before AFAIK). I have open a separate ticket here https://github.com/openzim/mwoffliner/issues/605

**Comment by kelson42** [comment]
I close this ticket as it looks there is nothing left to do. @Jaifroid @mossroy Thank you very much for your respective feedbacks. It was really helpful.

**Comment by mossroy** [comment]
The last ZIM files I could test still had some articles without any title.
It's not a blocker, but my opinion is that it would be cleaner to have a title for every article.

**Comment by kelson42** [comment]
@mossroy Do you have an example?

**Comment by kelson42** [comment]
@ISNIT0 Of course only the HTML tag should be removed, not what is inside as text

**Comment by mossroy** [comment]
The same articles mentioned above (see https://github.com/kiwix/kiwix-js/issues/475#issuecomment-467106241) have empty titles (based on kiwix-js) in the last ZIM file I could test (the one mentioned in https://github.com/openzim/mwoffliner/pull/606#issue-256032746)

**Comment by ISNIT0** [comment]
@kelson42 Of course the text is not removed, the logic gets the `textContent` from the html (https://github.com/openzim/mwoffliner/pull/606/files#diff-f9a469579c2c80d21239ff3f38522ffdR125).
Those are redirects I think, not articles. I guess they should retain the articleId as a title so it can be typed in to search etc.

**Comment by kelson42** [comment]
I had a look again on this in details and everything seems to work fine now. The last problems have been fixed regarding the ``title stuff``.

**Comment by Jaifroid** [comment]
@ISNIT0 `Wikivoyage_en_all_novid_2019-04.zim` is displaying hte "missing titles" symptoms discussed in this issue again. See screenshot below from Kiwix JS. All "titles" in square brackets are auto-generated by Kiwix JS from the URL, because the title is blank in the `dirEntry`. I'll post this also in https://github.com/openzim/mwoffliner/issues/605 .

![image](https://user-images.githubusercontent.com/4304337/55887441-6db7bc00-5ba5-11e9-9d6d-202146f183e9.png)

**Comment by Jaifroid** [comment]
This issue should be resolved with #488.

**Comment by Jaifroid** [comment]
Closed with #488.
